File: tradeit-android-sdk/src/main/java/it/trade/android/sdk/model/TradeItBalanceParcelable.java
package it.trade.android.sdk.model;


import android.os.Parcel;
import android.os.Parcelable;

import it.trade.model.reponse.TradeItAccountOverview;

public class TradeItBalanceParcelable implements Parcelable {

    public Double availableCash;

    public Double buyingPower;

    public Double dayAbsoluteReturn;

    public Double dayPercentReturn;

    public Double totalAbsoluteReturn;

    public Double totalPercentReturn;

    public Double totalValue;

    TradeItBalanceParcelable() {}

    TradeItBalanceParcelable(TradeItAccountOverview accountOverview) {
        this.availableCash = accountOverview.availableCash;
        this.buyingPower = accountOverview.buyingPower;
        this.dayAbsoluteReturn = accountOverview.dayAbsoluteReturn;
        this.dayPercentReturn = accountOverview.dayPercentReturn;
        this.totalAbsoluteReturn = accountOverview.totalAbsoluteReturn;
        this.totalPercentReturn = accountOverview.totalPercentReturn;
        this.totalValue = accountOverview.totalValue;
    }

    @Override
    public String toString() {
        return "TradeItBalanceParcelable{" +
                "availableCash=" + availableCash +
                ", buyingPower=" + buyingPower +
                ", dayAbsoluteReturn=" + dayAbsoluteReturn +
                ", dayPercentReturn=" + dayPercentReturn +
                ", totalAbsoluteReturn=" + totalAbsoluteReturn +
                ", totalPercentReturn=" + totalPercentReturn +
                ", totalValue=" + totalValue +
                '}';
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        TradeItBalanceParcelable that = (TradeItBalanceParcelable) o;

        if (availableCash != null ? !availableCash.equals(that.availableCash) : that.availableCash != null)
            return false;
        if (buyingPower != null ? !buyingPower.equals(that.buyingPower) : that.buyingPower != null)
            return false;
        if (dayAbsoluteReturn != null ? !dayAbsoluteReturn.equals(that.dayAbsoluteReturn) : that.dayAbsoluteReturn != null)
            return false;
        if (dayPercentReturn != null ? !dayPercentReturn.equals(that.dayPercentReturn) : that.dayPercentReturn != null)
            return false;
        if (totalAbsoluteReturn != null ? !totalAbsoluteReturn.equals(that.totalAbsoluteReturn) : that.totalAbsoluteReturn != null)
            return false;
        if (totalPercentReturn != null ? !totalPercentReturn.equals(that.totalPercentReturn) : that.totalPercentReturn != null)
            return false;
        return totalValue != null ? totalValue.equals(that.totalValue) : that.totalValue == null;

    }

    @Override
    public int hashCode() {
        int result = availableCash != null ? availableCash.hashCode() : 0;
        result = 31 * result + (buyingPower != null ? buyingPower.hashCode() : 0);
        result = 31 * result + (dayAbsoluteReturn != null ? dayAbsoluteReturn.hashCode() : 0);
        result = 31 * result + (dayPercentReturn != null ? dayPercentReturn.hashCode() : 0);
        result = 31 * result + (totalAbsoluteReturn != null ? totalAbsoluteReturn.hashCode() : 0);
        result = 31 * result + (totalPercentReturn != null ? totalPercentReturn.hashCode() : 0);
        result = 31 * result + (totalValue != null ? totalValue.hashCode() : 0);
        return result;
    }


    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeValue(this.availableCash);
        dest.writeValue(this.buyingPower);
        dest.writeValue(this.dayAbsoluteReturn);
        dest.writeValue(this.dayPercentReturn);
        dest.writeValue(this.totalAbsoluteReturn);
        dest.writeValue(this.totalPercentReturn);
        dest.writeValue(this.totalValue);
    }

    protected TradeItBalanceParcelable(Parcel in) {
        this.availableCash = (Double) in.readValue(Double.class.getClassLoader());
        this.buyingPower = (Double) in.readValue(Double.class.getClassLoader());
        this.dayAbsoluteReturn = (Double) in.readValue(Double.class.getClassLoader());
        this.dayPercentReturn = (Double) in.readValue(Double.class.getClassLoader());
        this.totalAbsoluteReturn = (Double) in.readValue(Double.class.getClassLoader());
        this.totalPercentReturn = (Double) in.readValue(Double.class.getClassLoader());
        this.totalValue = (Double) in.readValue(Double.class.getClassLoader());
    }

    public static final Parcelable.Creator<TradeItBalanceParcelable> CREATOR = new Parcelable.Creator<TradeItBalanceParcelable>() {
        @Override
        public TradeItBalanceParcelable createFromParcel(Parcel source) {
            return new TradeItBalanceParcelable(source);
        }

        @Override
        public TradeItBalanceParcelable[] newArray(int size) {
            return new TradeItBalanceParcelable[size];
        }
    };
}

File: tradeit-android-sdk/src/main/kotlin/it/trade/android/sdk/model/TradeItBalanceParcelable.kt
package it.trade.android.sdk.model


import android.os.Parcel
import android.os.Parcelable

import com.google.gson.annotations.SerializedName

import it.trade.model.reponse.TradeItAccountOverview

class TradeItBalanceParcelable : Parcelable {
    @SerializedName("availableCash")
    var availableCash: Double? = null

    @SerializedName("buyingPower")
    var buyingPower: Double? = null

    @SerializedName("dayAbsoluteReturn")
    var dayAbsoluteReturn: Double? = null

    @SerializedName("dayPercentReturn")
    var dayPercentReturn: Double? = null

    @SerializedName("totalAbsoluteReturn")
    var totalAbsoluteReturn: Double? = null

    @SerializedName("totalPercentReturn")
    var totalPercentReturn: Double? = null

    @SerializedName("totalValue")
    var totalValue: Double? = null

    @SerializedName("buyingPowerLabel")
    var buyingPowerLabel: String? = null

    @SerializedName("accountBaseCurrency")
    var accountBaseCurrency: String? = "USD"

    @SerializedName("marginCash")
    var marginCash: Double? = null

    internal constructor() {}

    internal constructor(accountOverview: TradeItAccountOverview) {
        this.availableCash = accountOverview.availableCash
        this.buyingPower = accountOverview.buyingPower
        this.dayAbsoluteReturn = accountOverview.dayAbsoluteReturn
        this.dayPercentReturn = accountOverview.dayPercentReturn
        this.totalAbsoluteReturn = accountOverview.totalAbsoluteReturn
        this.totalPercentReturn = accountOverview.totalPercentReturn
        this.totalValue = accountOverview.totalValue
        this.buyingPowerLabel = accountOverview.buyingPowerLabel
        this.accountBaseCurrency = accountOverview.accountBaseCurrency
        this.marginCash = accountOverview.marginCash
    }

    override fun toString(): String {
        return "TradeItBalanceParcelable{" +
                "availableCash=" + availableCash +
                ", buyingPower=" + buyingPower +
                ", dayAbsoluteReturn=" + dayAbsoluteReturn +
                ", dayPercentReturn=" + dayPercentReturn +
                ", totalAbsoluteReturn=" + totalAbsoluteReturn +
                ", totalPercentReturn=" + totalPercentReturn +
                ", totalValue=" + totalValue +
                ", buyingPowerLabel=" + buyingPowerLabel +
                ", accountBaseCurrency=" + accountBaseCurrency +
                ", marginCash=" + marginCash +
                '}'.toString()
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as TradeItBalanceParcelable

        if (availableCash != other.availableCash) return false
        if (buyingPower != other.buyingPower) return false
        if (dayAbsoluteReturn != other.dayAbsoluteReturn) return false
        if (dayPercentReturn != other.dayPercentReturn) return false
        if (totalAbsoluteReturn != other.totalAbsoluteReturn) return false
        if (totalPercentReturn != other.totalPercentReturn) return false
        if (totalValue != other.totalValue) return false
        if (buyingPowerLabel != other.buyingPowerLabel) return false
        if (accountBaseCurrency != other.accountBaseCurrency) return false
        if (marginCash != other.marginCash) return false

        return true
    }

    override fun hashCode(): Int {
        var result = availableCash?.hashCode() ?: 0
        result = 31 * result + (buyingPower?.hashCode() ?: 0)
        result = 31 * result + (dayAbsoluteReturn?.hashCode() ?: 0)
        result = 31 * result + (dayPercentReturn?.hashCode() ?: 0)
        result = 31 * result + (totalAbsoluteReturn?.hashCode() ?: 0)
        result = 31 * result + (totalPercentReturn?.hashCode() ?: 0)
        result = 31 * result + (totalValue?.hashCode() ?: 0)
        result = 31 * result + (buyingPowerLabel?.hashCode() ?: 0)
        result = 31 * result + (accountBaseCurrency?.hashCode() ?: 0)
        result = 31 * result + (marginCash?.hashCode() ?: 0)
        return result
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeValue(this.availableCash)
        dest.writeValue(this.buyingPower)
        dest.writeValue(this.dayAbsoluteReturn)
        dest.writeValue(this.dayPercentReturn)
        dest.writeValue(this.totalAbsoluteReturn)
        dest.writeValue(this.totalPercentReturn)
        dest.writeValue(this.totalValue)
        dest.writeString(this.buyingPowerLabel)
        dest.writeString(this.accountBaseCurrency)
        dest.writeValue(this.marginCash)
    }

    protected constructor(`in`: Parcel) {
        this.availableCash = `in`.readValue(Double::class.java.getClassLoader()) as? Double
        this.buyingPower = `in`.readValue(Double::class.java.getClassLoader()) as? Double
        this.dayAbsoluteReturn = `in`.readValue(Double::class.java.getClassLoader()) as? Double
        this.dayPercentReturn = `in`.readValue(Double::class.java.getClassLoader()) as? Double
        this.totalAbsoluteReturn = `in`.readValue(Double::class.java.getClassLoader()) as? Double
        this.totalPercentReturn = `in`.readValue(Double::class.java.getClassLoader()) as? Double
        this.totalValue = `in`.readValue(Double::class.java.getClassLoader()) as? Double
        this.buyingPowerLabel = `in`.readString()
        this.accountBaseCurrency = `in`.readString()
        this.marginCash = `in`.readValue(Double::class.java.getClassLoader()) as? Double
    }

    companion object {
        @JvmField
        val CREATOR: Parcelable.Creator<TradeItBalanceParcelable> = object : Parcelable.Creator<TradeItBalanceParcelable> {
            override fun createFromParcel(source: Parcel): TradeItBalanceParcelable {
                return TradeItBalanceParcelable(source)
            }

            override fun newArray(size: Int): Array<TradeItBalanceParcelable?> {
                return arrayOfNulls(size)
            }
        }
    }
}
